Shorts tab on channel page

KarateCowboy opened the pull request:
# Add 'Shorts' tab to the Channel page



<!-- Thanks for sending a pull request! Make sure to follow the contributing guidelines. -->
<!-- Important note, we may remove your pull request if you do not use this provided PR template correctly. -->

## Pull Request Type
<!-- Please select what type of pull request this is: [x] -->
- [ ] Bugfix
- [x] Feature Implementation
- [ ] Documentation
- [ ] Other

## Description
Adds the Shorts tab for a channel to the Channel Page. Allows for playing via direct or invidious api

## Screenshots

### Before:
![before](https://user-images.githubusercontent.com/361120/217639040-77d1c1f3-a6e3-419e-8b15-f8e3b2feccbc.png)

#### After:

![after](https://user-images.githubusercontent.com/361120/217639103-f9241a32-bf65-403f-819b-fa5bf602f72b.png)


## Testing

1) Find a channel which has Shorts and Subscribe to it.
2) View the Channels page, then click on that channel
3) The new "Shorts" tab should be visible. Click it
4) There should be a grid with Shorts in it. 
5) Click any of the shorts. You should be taken to the Watch page and be able to watch the Short

absidue commented:
This PR doesn't actually implement the feature request that you mention in the Related issue section. As mentioned in https://github.com/FreeTubeApp/FreeTube/issues/2476#issuecomment-1216036544 that feature request is actually for the infinitely scrolling shorts tab, although there is an unrelated discussion towards the end of that thread that mentions the shorts tab on the channels page.
I'll remove it from the pull request description for you, so that it stays open until someone actually implements it.

absidue commented on package.json:
```diff
@@ -77,8 +77,8 @@
     "vuex": "^3.6.2",
-    "youtubei.js": "^2.8.0",
-    "yt-channel-info": "^3.2.1"
+    "youtubei.js": "^2.9.0",
```
I can't see any code in this pull request, that uses any of the new changes in YouTube.js 2.9.0, so it doesn't need to be upgraded as part of this PR.
```suggestion
    "youtubei.js": "^2.8.0",
```

absidue commented on src/renderer/components/ft-list-video/ft-list-video.js:
```diff
@@ -393,7 +393,7 @@ export default Vue.extend({
       }
 
-      if (typeof (this.data.publishedText) !== 'undefined' && this.data.publishedText !== null && !this.isLive) {
+      if ('publishedText' in this.data && !!this.data.publishedText && !this.isLive) {
```
This change can be simplified to this:
```suggestion
      if (this.data.publishedText && !this.isLive) {
```

KarateCowboy commented:
Hi,

So, in issue #2476 there is quite a bit going on. 

- The opening post by a deleted user asks for a dedicated shorts tab but is ambiguous beyond that. We obviously cannot ask for clarification.
- Then @ChunkyProgrammer [speculates](https://github.com/FreeTubeApp/FreeTube/issues/2476#issuecomment-1216036544) that the user _may_ have wanted an infinite scrolling list of some type.
- Then we see @efb4f5ff-1298-471a-8973-3d47447115dc [saying](https://github.com/FreeTubeApp/FreeTube/issues/2476#issuecomment-1304549134) there should be a shorts tab that is just like the subscriptions tab. All shorts, from all channels
- I thought there was also a user asking for a shorts tab on the channels page. Is that what you removed?
- Finally, there is @ChunkyProgrammer [commenting](https://github.com/FreeTubeApp/FreeTube/issues/2476#issuecomment-1304549134) that there should be a toggle to disable both shorts and streams


I see content for several stories in there:

1. A shorts tab on the channel page
2. a shorts page which includes aggregated shorts from all subscriptions
3. a streams tab on the channel page
4. an infinite scrolling loop of shorts with autoplay
5. the ability to disable shorts
6. the ability to disable streams

How about we use this PR as an MVP to get shorts, at all, out the door(item 1 on the list)? I can go and create separate stories for the rest, and consult others to get clarification for details about them. Let me know your thoughts, everyone.

absidue commented:
The only thing I edited was the description of this pull request, to remove the reference to a feature request for a different feature (See the screenshot below).

Why am I so sure that the original feature request is unrelated to the shorts tab on the channel page?
That feature request was opened in August, YouTube only split the videos tab up into the videos, shorts and live tabs in October/November. Additionally the requester asked for a dedicated shorts tab and Chunky pointed out that the YouTube has a dedicated infinitely scrollable shorts tab (visit https://www.youtube.com and click shorts in the sidebar), so it's highly likely that the requester meant that.

This pull request only implements the discussion that is going on below that feature request, not the actual feature request itself. I removed your "implements #number" text, so the feature request stays open until it is implemented.

I'm sorry that my words and actions came across as attacks, that wasn't my intention. My intention was just to point out that the feature that this pull request implements, is not the one in the feature request that you linked, so I removed your link to it. This pull request is completely fine (well a few code changes are needed before it gets merged, like fixing the code conflicts) as a pull request that implements the shorts tab on the channel page.

![pr-edit](https://user-images.githubusercontent.com/48293849/213876124-da89fac9-a838-422c-87bb-b55787442bfe.png)

KarateCowboy commented:
Hi absidue

Forgive me if the verbosity of my last post made it seem like I was being defensive. I interpret what you wrote as combative. There is much to this project which I am still learning: what code does what, who is active and coding, and who is keeping track of what. Because of that I decided to be very detailed about what I observe and am thinking.  Also, I am excited and charged to help out, so there is that energy.

What you are saying about Youtube splitting pieces versus the creation of the feature request makes perfect sense. I totally agree.

I will clean up the code conflicts. There are two outstanding things I see where I need some advise:

1. What do I need to do about i18n for the features, if anything?
2. My code is pointing to a git commit, per @ChunkyProgrammer 's advice. Do I need to wait for a new release, or is there a specific version I should point to now?


After that, what is the review, test, and release process?

ChunkyProgrammer commented on src/renderer/views/Channel/Channel.js:
```diff
@@ -509,6 +542,34 @@ export default Vue.extend({
+        resource: `channels/${this.id}/shorts`,
+        id: '',
+        params: {
+          sort_by: this.playlistSortBy
```
This is not needed for shorts

KarateCowboy replied:
✅

ChunkyProgrammer commented on src/renderer/views/Channel/Channel.js:
```diff
@@ -509,6 +542,34 @@ export default Vue.extend({
+      }
+      this.invidiousAPICall(payload).then((response) => {
+        this.latestShorts = response.videos
+        this.videoContinuationString = response.continuation
```
Should probably add a separate variable for storing the shorts continuation

KarateCowboy replied:
✅

ChunkyProgrammer commented on src/renderer/views/Channel/Channel.js:
```diff
@@ -509,6 +542,34 @@ export default Vue.extend({
+        })
+        if (process.env.IS_ELECTRON && this.backendPreference === 'invidious' && this.backendFallback) {
+          showToast(this.$t('Falling back to Local API'))
+          this.getPlaylistsLocal()
```
should call getShortsLocal instead

KarateCowboy replied:
✅

ChunkyProgrammer commented on src/renderer/views/Channel/Channel.js:
```diff
@@ -352,6 +360,31 @@ export default Vue.extend({
+        })
+        if (this.backendPreference === 'local' && this.backendFallback) {
+          showToast(this.$t('Falling back to Invidious API'))
+          this.getChannelInfoInvidious()
```
should call getShortsInvidious

KarateCowboy replied:
✅

ChunkyProgrammer commented on src/renderer/views/Channel/Channel.js:
```diff
@@ -352,6 +360,31 @@ export default Vue.extend({
+        }
+
+        this.latestShorts = response.items
+        this.videoContinuationString = response.continuation
```
Should probably add a separate variable for storing the shorts continuation

KarateCowboy replied:
✅

ChunkyProgrammer commented on src/renderer/views/Watch/Watch.js:
```diff
@@ -626,7 +626,7 @@ export default Vue.extend({
           }
 
-          if (result.storyboards?.type === 'PlayerStoryboardSpec') {
+          if (result.storyboards?.type === 'PlayerStoryboardSpec' && result.storyboards.boards.length >= 3) {
```
Did you get errors from this code so you added this check or did you want to make the if statement safer?

KarateCowboy replied:
I got undefined errors. Shorts I was testing do not have any storyboards. So, the next line was passing undefined to `createStoryboardUrls`. So 'yes' to both.

absidue replied:
Could you provide some of the links of the shorts that were erroring, the ones I tested (Linus Tech Tips YouTube channel) all worked, so I would like to see if they have other storyboards.

PikachuEXE replied:
I already don't understand the `result.storyboards?.type === 'PlayerStoryboardSpec'` here
With the new `result.storyboards.boards.length >= 3` it's even harder to understand, add some code comments?

PikachuEXE replied:
I missed `result.storyboards.boards[2]` below oops
I find this project lacks code comments very much 😬

ChunkyProgrammer commented on static/locales/en-US.yaml:
```diff
@@ -512,6 +512,8 @@ Channel:
     results
   Sort By: Sort By
+  Shorts:
+    Shorts: Shorts
```
Add `This channel does not currently have any shorts` below

KarateCowboy replied:
✅

ChunkyProgrammer reviewed (commented):
1) You are missing a string for 'This channel does not currently have any shorts' in the en-US locale file. Other than that i18n looks good.
![image](https://user-images.githubusercontent.com/78101139/214175224-d133639a-9d0f-4047-9d00-2da2ecc83b1c.png)

The translations for other languages are done through Weblate and if a translation is missing in a language, it will fallback to en-US 😄 

2) Pointing to the git branch should be fine for now (we'll be using YouTube.js in FreeTube's next release instead of yt-channel-info)

"Also, while the request mentioned the ability to hide shorts, I could find no such toggle. Perhaps I could add that on to this." sorry, I just didnt remember off the top of my head if that functionality was already implemented or not 😄 if you would like to add it, you can do it in a separate PR.

If you'd like to add the ability to load more shorts to this PR:
a continuation parameter can be added to the invidious call or you can call getChannelVideosMore with the shorts continuation 

Let me know if you have any more questions. Thanks for opening a PR!

KarateCowboy commented:
I updated and took all of @ChunkyProgrammer 's suggestions. It appears that additional shorts are automatically loading as I scroll down the page. Hopefully it works the same for you.

> Thanks for opening a PR!

Thank you for creating this application. And, thank you for facilitating my contribution.

ChunkyProgrammer commented:
Would you be able to resolve the merge conflict as well?

KarateCowboy commented:
@ChunkyProgrammer Sure thing. It looks like I missed the part where you said it is OK to point to the git commit.

@absidue Sure thing. To replicate the error, remove the extra conditional. Then subscribe to the channel [Julia Stunts](https://www.youtube.com/channel/UCyViBfYJKESvPOIFRBT2ymg). Then go to that channel via "Channels > Julia Maggio". Then click the shorts tab. Then the Short "Devil Hunters@WholeWheatPete @itsKingChris & @StellaChuu".

absidue commented:
@KarateCowboy Looks like YouTube returns different storyboard data for those very short shorts, only a single batch of storyboards and they look like they need parsing differently. So for the moment your "ignore the storyboards entirely" change is alright as it means that it doesn't error.
I'll create a separate pull request to fix them properly, as I still need to work out how to handle that different storyboard data.

PikachuEXE commented on src/renderer/views/Channel/Channel.js:
```diff
@@ -508,6 +542,32 @@ export default defineComponent({
+        } else {
+          this.isLoading = false
+        }
+      }
+      )
```
This looks really strange (even it's valid)
```suggestion
      })
```

KarateCowboy replied:
I don't know the significance of that line selecting the storyboard at index = 2 because I am new to the project. I just know that when given a Short the function was receiving undefined. I would not have touched it if it were not throwing errors.

PikachuEXE commented:
What do I do to test this?
Coz I have no idea what's the intended behaviour
Please update description so any new comer can see that too

PikachuEXE reviewed (dismissed):
Please also test channels **without** shorts

Tested channels with shorts
- https://www.youtube.com/channel/UCXuqSBlHAE6Xw-yeJA0Tunw
- https://www.youtube.com/channel/UCajiMK_CY9icRhLepS8_3ug

![image](https://user-images.githubusercontent.com/1018543/215382685-662c911c-d570-4936-8338-f1fd5bb849bf.png)


Tested channel **without** shorts
- https://www.youtube.com/channel/UCz7eke4JGlbtc11_6mmA7ew

![image](https://user-images.githubusercontent.com/1018543/215382607-5efb44f5-933c-41e5-bbdd-85a81c9c8089.png)

ChunkyProgrammer commented on src/renderer/views/Channel/Channel.js:
```diff
@@ -505,6 +539,31 @@ export default defineComponent({
+        id: '',
+        params: {}
+      }
+      this.invidiousAPICall(payload).then((response) => {
```
Remove `this.`
```suggestion
      invidiousAPICall(payload).then((response) => {
```

KarateCowboy replied:
committed

absidue reviewed (commented):
Why are the shorts tied to the video sort by?

KarateCowboy commented:
> Why are the shorts tied to the video sort by?

At the time I was thinking the functionality is the same on the Shorts page as the Videos page. I'll make sure the shorts page has its own method and strip apart if needed.

KarateCowboy commented:
Also going to remove some of the grossness I found in my additions.

KarateCowboy commented:
> This pull request has conflicts, please resolve those before we can evaluate the pull request.

👍🏻 still under construction. Was working on it last night 🚧

KarateCowboy commented:
Okey dokey, so:

I went out on a limb and added Jest so I can write tests before I write stuff. Personally I prefer the old Mocha/Chai/Sinon combo. But, everybody and their mother supports Jest these days _because Facebook__. Perhaps more pertinent: Vue3/Vite/Vue-test-utils all have good support for it. I saw in one of project pages that @ChunkyProgrammer made cards about upgrading to Vue3 and adding TypeScript. Having a good test suite will make that much easier.

The one caveat is that jest cannot mock/spyon function calls from outside a module IF the function is then called within the module.  See [here](https://medium.com/welldone-software/jest-how-to-mock-a-function-call-inside-a-module-21c05c57a39f). Hence, I brought in Sinon to do the job.

The Channeljs page has many methods which are the same except swapping playlists/videos/shortages or YT/Invidious.  I farmed the API calls for YT/Invidious out to the api helper library to observe separation of concerns. This will be useful when creating a dedicated Channels page from the left  menu, or when creating an endless scrolling Shorts page.

As it stands the Invidious api supports continuation for shorts but the YT channel library does not, so I left that. I implemented the Sorting, but it does not seem to work because of something with the API libraries.

Let me know your thoughts.

KarateCowboy commented:
> Why are the shorts tied to the video sort by?

My mistake. Fixed in this latest commit.

KarateCowboy commented:
Okey dokey, so:

I went out on a limb and added Jest so I can write tests before I write stuff. Personally I prefer the old Mocha/Chai/Sinon combo. But, everybody and their mother supports Jest these days because Facebook_. Perhaps more pertinent: Vue3/Vite/Vue-test-utils all have good support for it. I saw in one of project pages that @ChunkyProgrammer made cards about upgrading to Vue3 and adding TypeScript. Having a good test suite will make that much easier.

The one caveat is that jest cannot mock/spyon function calls from outside a module IF the function is then called within the module. See [here](https://medium.com/welldone-software/jest-how-to-mock-a-function-call-inside-a-module-21c05c57a39f). Hence, I brought in Sinon to do the job.

The Channeljs page has many methods which are the same except swapping playlists/videos/shortages or YT/Invidious. I farmed the API calls for YT/Invidious out to the api helper library to observe separation of concerns. This will be useful when creating a dedicated Channels page from the left menu, or when creating an endless scrolling Shorts page.

As it stands the Invidious api supports continuation for shorts but the YT channel library does not, so I left that. I implemented the Sorting, but it does not seem to work because of something with the API libraries.

Let me know your thoughts.

PS: Sorry for the whitespace changes. It looks like some functions have a space between the name and first paren, but some do not.

absidue commented:
Hi
I'm definitely not against adding tests, however please do it in a separate pull request, I'm not a fan of having one pull request that does many unrelated changes, because that makes it extremely difficult to track which commit add/removed what.
I know it's your first pull request to FreeTube but may I remind you that you are allowed to do multiple, you don't have to cram all your contributions into a single pull request :).

absidue commented on src/renderer/helpers/api/index.js:
```diff
@@ -0,0 +1,8 @@
+import * as invidious from './invidious'
+import * as local from './local'
+
+export const channelShorts = async (api, channelId, idType, sortBy) => {
+  const shortsFunc = api === 'invidious' ? invidious.channelShortsInvidious : local.channelShortsLocal
+  const result = await shortsFunc(channelId, idType, sortBy)
+  return api === 'invidious' ? invidious.parseShortsResponse(result) : local.parseShortsResponse(result)
+}
```
I'm also not a fan of this change, as it will have to be reverted when #3143 is merged as YouTube.js works differently to yt-channel-info.

KarateCowboy commented:
Okey dokey.... MUCH better.

No more of that noisy whitespace changing stuff.

The API calls for shorts did not exist prior to this PR, so it's not a refactor to put them in the API helpers file. 

If you like, I can pull the test & test boiler plate into a different PR. 

Please take a fresh look when you have time. It's much cleaner now.

efb4f5ff-1298-471a-8973-3d47447115dc commented:
@KarateCowboy could u please upload the screenshots direct on GitHub instead of Dropbox

KarateCowboy commented:
> @KarateCowboy could u please upload the screenshots direct on GitHub instead of Dropbox

Done

efb4f5ff-1298-471a-8973-3d47447115dc commented:
Noticed that the view count is off

KarateCowboy commented:
> Noticed that the view count is off

I looked and that's actually what is coming through from yt-channel-info. if you `console.log` the response in `src/renderer/helpers/api/index.js` you can see:

![console_and_view](https://user-images.githubusercontent.com/361120/218019523-8d6d6cc7-e2e0-4071-8ff7-9566b030f653.png)

And the same Short on YT:
![yt_source](https://user-images.githubusercontent.com/361120/218019560-b20e3323-0b0e-4429-aca3-fea50c14c853.png)

absidue commented:
Could you please add the new tab to the tabInfoValues array on the channel page. That array is used so you can cycle the focus through the tabs by pressing the left and right arrow keys.

ChunkyProgrammer commented:
> PS: This YouTube.js stuff --- how close to arrival is it?

Once https://github.com/LuanRT/YouTube.js/pull/309 gets merged then https://github.com/FreeTubeApp/FreeTube/pull/3143 will need some updates and then it can be merged

KarateCowboy commented:
> > PS: This YouTube.js stuff --- how close to arrival is it?
> 
> The current plan: Once [LuanRT/YouTube.js#309](https://github.com/LuanRT/YouTube.js/pull/309) gets merged then #3143 will need some updates + testing + review and then the youtubei.js migration will be merged when approved.

Understood. I will keep an eye on that and then update accordingly.

efb4f5ff-1298-471a-8973-3d47447115dc commented:
@KarateCowboy #3143 is merged!

KarateCowboy commented:
> @KarateCowboy #3143 is merged :)

I'll get to it this weekend!

efb4f5ff-1298-471a-8973-3d47447115dc commented:
> > @KarateCowboy #3143 is merged :)
> 
> I'll get to it this weekend!

umm this weekend @KarateCowboy ?

KarateCowboy commented:
Sorry. I looked at it a little last weekend but had my time cut short because I started a new job. This weekend, however, I have more time. Even if it's my birthday. :D

Mar 11, 2023 01:11:53 efb4f5ff-1298-471a-8973-3d47447115dc ***@***.***>:

> 
> @KarateCowboy[https://github.com/KarateCowboy] #3143[https://github.com/FreeTubeApp/FreeTube/pull/3143] is merged :)
> 
> I'll get to it this weekend!
> 
> umm this weekend @KarateCowboy[https://github.com/KarateCowboy] ?
> 
> —
> Reply to this email directly, view it on GitHub[https://github.com/FreeTubeApp/FreeTube/pull/3093#issuecomment-1464847994], or unsubscribe[https://github.com/notifications/unsubscribe-auth/AACYFICI4RB2P44KJR65DS3W3QQTDANCNFSM6AAAAAAT7RTW3U].
> You are receiving this because you were mentioned.[Tracking image][https://github.com/notifications/beacon/AACYFIHHSSISB5YJ3GHSQ43W3QQTDA5CNFSM6AAAAAAT7RTW3WWGG33NNVSW45C7OR4XAZNMJFZXG5LFINXW23LFNZ2KUY3PNVWWK3TUL5UWJTSXJ7HHU.gif]
>

absidue commented:
@KarateCowboy feel free to use #3273 as an inspiration (it adds support for the live tab).

KarateCowboy commented:
pulled in latest. hacking away a little tonight

ChunkyProgrammer commented:
Hi @KarateCowboy do you still plan on working on this? Note: invidious shorts should work once https://github.com/iv-org/invidious/pull/3700 is merged

KarateCowboy commented:
> Hi @KarateCowboy do you still plan on working on this? Note: invidious shorts should work once [iv-org/invidious#3700](https://github.com/iv-org/invidious/pull/3700) is merged

Hey hey. Thanks for the bump. I'll hit it up over the holiday weekend. Sunday night

absidue commented:
Hi @KarateCowboy
Glad to hear that you are picking this up again :). To keep this pull request small and on topic, please leave the addition of a test framework and the refactor out of this pull request. If you want to add it, please do so in a different pull request, instead of cramming multiple things into one pull request. Keeping it simple, makes it easier and faster to review.

efb4f5ff-1298-471a-8973-3d47447115dc commented:
Closing, this is implemented in https://github.com/FreeTubeApp/FreeTube/pull/3533
